src/coreComponents/constitutive/ConstitutiveManager.cpp

```diff
@@ -55,33 +55,59 @@ void ConstitutiveManager::expandObjectCatalogs()
 }
 
 
-ConstitutiveBase &
+void
 ConstitutiveManager::hangConstitutiveRelation( string const & constitutiveRelationInstanceName,
                                                dataRepository::Group * const parent,
                                                localIndex const numConstitutivePointsPerParentIndex ) const
 {
-  ConstitutiveBase const & constitutiveRelation = getConstitutiveRelation( constitutiveRelationInstanceName );
-
-  std::unique_ptr< ConstitutiveBase > material = constitutiveRelation.deliverClone( constitutiveRelationInstanceName, parent );
-
-  material->allocateConstitutiveData( *parent,
-                                      numConstitutivePointsPerParentIndex );
-
   dataRepository::Group * constitutiveGroup = parent->getGroupPointer( groupKeyStruct::constitutiveModelsString() );
-  if( constitutiveGroup == nullptr )
+  if( constitutiveGroup == nullptr )   // Why do we need to do this check? shouldn't it always be there?
   {
     constitutiveGroup = &parent->registerGroup( groupKeyStruct::constitutiveModelsString() ).setSizedFromParent( 1 );
     constitutiveGroup->resize( parent->size() );
   }
 
-
-  ConstitutiveBase &
-  rval = constitutiveGroup->registerGroup< ConstitutiveBase >( constitutiveRelationInstanceName, std::move( material ) );
-  rval.setSizedFromParent( 1 );
-  rval.resize( constitutiveGroup->size() );
-  return rval;
+  // 1. Allocate constitutive relation
+  // we only register the group if the instance has not been registered yet.
+  if( !constitutiveGroup->hasGroup( constitutiveRelationInstanceName ) )
+  {
+    ConstitutiveBase const & constitutiveRelation = getConstitutiveRelation( constitutiveRelationInstanceName );
+
+    std::unique_ptr< ConstitutiveBase > material = constitutiveRelation.deliverClone( constitutiveRelationInstanceName, parent );
+
+    material->allocateConstitutiveData( *parent,
+                                        numConstitutivePointsPerParentIndex );
+
+    ConstitutiveBase &
+    rval = constitutiveGroup->registerGroup< ConstitutiveBase >( constitutiveRelationInstanceName, std::move( material ) );
+    rval.setSizedFromParent( 1 );
+    rval.resize( constitutiveGroup->size() );
+
+    // 2. Allocate subrelations (for compound models)
+    std::vector< string > const subRelationNames = constitutiveRelation.getSubRelationNames();
+    for( string const & subRelationName : subRelationNames )
+    {
+      std::cout << "nothing to do" << std::endl;
+      ConstitutiveBase const & subRelation = getConstitutiveRelation( subRelationName );
+
+      std::unique_ptr< ConstitutiveBase > constitutiveModel = subRelation.deliverClone( subRelationName, parent );
+
+      constitutiveModel->allocateConstitutiveData( *parent,
+                                                   numConstitutivePointsPerParentIndex );
+
+      // we only register the group if the instance has not been registered yet.
+      if( !constitutiveGroup->hasGroup( subRelationName ) )
```

Comment:
I am not 100% certain, but if the sub-relation has already been registered (presumably by another compound model, or explicitly by the user by listing it in region's material list), should we also avoid calling `deliverClone` and `allocateConstitutiveData` above, i.e. move entire loop body under this branch?

Reply:
yes, I think you are right. I don't see why we would not want to do it.